### PR TITLE
Exposes sequence numbers for all requests and responses

### PR DESCRIPTION
**Pull request by hidmic**
This pull request exposes sequence numbers for every action related service request and response to enable 1-to-1 correlation between them (on top of goal ID based correlations that may exist between requests and responses).

### CODE REVIEW

**Comment by apojomovsky** [comment]
CI status
Linux: [Build Status](https://ci.ros2.org/job/ci_linux/5615/)
Linux-aarch64: [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/2242/)
OSX: [Build Status](https://ci.ros2.org/job/ci_osx/4624/)
Windows: [Build Status](https://ci.ros2.org/job/ci_windows/5491/) (waiting for executor)

File: rcl_action/include/rcl_action/action_client.h [hunk 1]
```diff
@@ -274,7 +275,8 @@ RCL_WARN_UNUSED
 rcl_ret_t
 rcl_action_send_goal_request(
   const rcl_action_client_t * action_client,
-  const void * ros_goal_request);
+  const void * ros_goal_request,
+  int64_t * sequence_number);
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rcl_action/include/rcl_action/action_client.h]: @jacobperron the reason why I sent this PR in the first place was to simplify correlating promises in the upper layer. That's necessary for cancellation requests (e.g. you may have concurrent cancel requests for a given goal ID through the goal handle and for multiple goal IDs by not specifying them). 

For goal requests you can get away with just the goal ID if we keep that mapping on the server and throw from the client if the same goal ID is sent again (it starts getting tricky though, we need expiration for that tracking collection as well). For result requests the same applies, though I'm not sure if we want to error if multiple concurrent result requests are sent. I guess we could return an existing promise if another result request gets in while there's one in course already.

All in all, I think that having unique identifiers for each transaction may be helpful. But I agree we should come back once `rclcpp` and `rclpy` have been implemented to remove any redundancies.

**Comment by hidmic** [comment]
CI is having issues :(

**Comment by hidmic** [comment]
@sloretz could you re-trigger CI? (assuming it is now working)

**Comment by sloretz** [comment]
Windows [Build Status](https://ci.ros2.org/job/ci_windows/5498/)

**Comment by sloretz** [comment]
OSX (using CI args to test just rcl_action)
[Build Status](https://ci.ros2.org/job/ci_osx/4643/)

**Pull request merged**